Issue #585: allow users to specify columns and forecast unit in `as_forecast()`

nikosbosse opened the pull request:
<!-- Thanks for opening this pull request! Below we have provided a suggested template for PRs to this repository and a checklist to complete before opening a PR -->
 
## Description

This PR closes #585 

As discussed in #585 it is convenient for users to be able to specify the forecast unit, as well as changes to column names they need to make as part of `as_forecast()`. 

This PR
- adds additional arguments `observed`, `predicted`, `model`, `forecast_unit`, `quantile_level`, `sample_id` to `as_forecast()` that allow the users to specify the desired forecast unit as well as desired changes to column names
- adds checks to validate the inputs to `as_forecast()`
- adds tests to check the behaviour is as expected
- updates the NEWS file

Additional thoughts and considerations: 
- the order of the arguments could be different. For example, we might want `forecast_unit` to be the first argument. Strong opinions? At the moment I put it there because I felt it was more natural to first specify the columns to be renamed and then the forecast unit and then the special columns. But 🤷 
- We could in principle create extra methods for sample and quantile-based forecasts (and then move the arguments `sample_id` and `quantile_level` to those methods. As mentioned in #585 I feel this would lead to unnecessary complexity (having to call `as_forecast()` --> `as_forecast.default()` --> `as_forecast()` --> `as_forecast.forecast_sample()` just to hide an argument that is clearly explained in the docs). 

## Checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have included the target issue or issues in the PR title as follows: *issue-number*: PR title
- [x] I have tested my changes locally.
- [x] I have added or updated unit tests where necessary.
- [x] I have updated the documentation if required.
- [x] I have built the package locally and run rebuilt docs using roxygen2.
- [x] My code follows the established coding standards and I have run `lintr::lint_package()` to check for style issues introduced by my changes. 
- [x] I have added a news item linked to this PR.
- [x] I have reviewed CI checks for this PR and addressed them as far as I am able.

<!-- Thanks again for this PR - @scoringutils dev team -->

toshiakiasakura commented:
For me, giving a partial argument like `as_forecast(data, observed="value")` seems weird to me. Personally, I regard the as_forecast as creating a new class object like the S4 object method allowing the flexibility of using data.table operation with the S3 method. 

So, I prefer to expect the following behaviour. If `observed`, `predicted` and `model` are prepared, `as_forecast(data)` works. Otherwise, we need to specify every argument (`quantile_level` and `sample_id` as an optional argument). Then, users experience errors if they fail to specify any of the required arguments. Also, it does not break the existing code base. 

The documentation looks like this. 
```R
     as_forecast(data, ...)
     
     ## S3 method for class 'forecast'
     as_forecast(data, ...)
     
     ## Default S3 method:
     as_forecast(
       data,
       observed,
       predicted,
       model,
       forecast_unit,
       quantile_level = NULL,
       sample_id = NULL,
       ...
     )
```

However, this type of dispatch is not supported in the S3 method, so my idea is to introduce the meta-class `forecast` before assigning a specific `forecast` object. (Lightly discussed in #592). We can achieve the above functionality by writing in the following way. 

```R
as_forecast <- function(data,
                        ...) {
  flag <- c("observed", "predicted", "model") %in% names(data)
  if (all(flag) == TRUE) {
      data <- ensure_data.table(data)
      class(data) <- c("forecast", class(data))
  }
  UseMethod("as_forecast", data)
}
# as appropriate
as_forecast.forecast <- function()
as_forecast.default <- function()
```

What do you think about this approach?

nikosbosse commented:
@toshiakiasakura Thanks for your comment. I'm not sure I understand everything correctly, so let me try and summarise it. What I understood is that you want to 
- first transform the object into an object of class `forecast` and then
- transform the object again into an object of class `forecast_binary`, `forecast_point`, etc.

And you would do this so that users can either 
- use `as_forecast()` without any additional arguments if there are already columns `observed`, `predicted`, and `model` (what happens with the argument `forecast_unit` here?)
- or such that they are forced to provide all arguments (even if they, for example, already might have a column called `model` and one called `predicted`)

I'm not sure what the benefit is of forcing users to provide all arguments even if they don't need all of them. I also think it's confusing that sometimes you have to provide all arguments and sometimes you don't have to provide anything.
Also `forecast_unit` would be an optional argument in both cases, right? 

> Also, the advantage of this approach is that this change does not break the existing code base.

The current proposal also wouldn't break the existing code base. 

I can see why you would force users to always provide all arguments. But since that's not necessary in many instances it would make things more cumbersome to users and would definitely break existing code. 

Again not sure I understand correctly. But it seems like this proposal would introduce a lot of complexity both in terms of what the user has to do and in terms of code complexity.

toshiakiasakura commented:
Yes, I think you understand what I want to say mostly. If I add something, 
- The need to transform the object into a class `forecast` comes from wanting to apply the multiple dispatch for argument specification (no arguments v.s. set of arguments). So, essentially the `forecast` class can be removed from the returned object (but the documentation will remain). Currently I am thinking to introduce `forecast` class also, meaning that the returned object has a class like `forecast_binary, forecast, data.table, data.frame`, but this probably needs further discussion. 
- In my suggestion, the purpose of `as_forecast()` is for users who know what arguments are required in scoringutils and most introductory users go for `as_forecast(data, observed, predicted, model)` type of specification. 

I think the benefit of forcing users is the explicitness of the function usage. In my opinion, the current behaviour of `as_forecast()` without providing any arguments, which in fact uses pre-specified column names, seems to me like applying a linear regression model with `lm(data=data)` which implicitly uses `outcome` and `explanatory` columns as an outcome and explanatory variable. So, for me, using `protected` columns implicitly is a confusing point. So, I prefer to prepare first the full specification type of `as_forecast` and second I notice that `as_forecast()` is also required (not to break the existing code set and probably a large-scale usage), which results in my current suggestion. 

Probably I am missing the following points. 
- I am not sure of the real-world application of scoringutils. My suggestion comes from my small toy experience with this package.  Do you know a publicly available in-built use case of this package? 
- Even in large-scale package development, I guess the use of partial specification of arguments like `as_forecast(data, observed="value")` would be rare since an experienced developer would try to adjust to use `as_forecast()`, and introductory users would use a full specification. 

> Also forecast_unit would be an optional argument in both cases, right?

Yes right, but if you intend to let users specify `forecast_unit` at this stage (and in terms of my preference of the explicitness), I will leave it as a required argument, and prepare `c("a", "b", "c")` to set specific columns, `TRUE` to set all the remaining columns, and `FALSE` not to set a forecast unit at this stage.  

 > The current proposal also wouldn't break the existing code base.

Sorry, you are true. I just compare the approach against dropping `as_forecast()` version from the implementation, which breaks the existing code.

nikosbosse commented:
I think the simpler version of this would be to make all arguments mandatory (including `forecast_unit`) and not worry about breaking things (since we're introducing all kinds of breaking changes anyway). But as a user, I think I wouldn't like this as it forces me to do a lot of extra typing even if I know what I'm doing. Giving the user an option is nice, but personally I wouldn't force them. 

@seabbs @sbfnk do you have any thoughts on this?

toshiakiasakura commented:
Another choice would be 
- making all arguments mandatory for `as_forecast` and preparing another function like `simple_as_forecast` for the current behaviour of `as_forecast()` (without any arguments) for an experienced user. 

Essentially this is the same as I suggested, but it is more natural in terms of the implementation.

seabbs commented:
> But as a user, I think I wouldn't like this as it forces me to do a lot of extra typing even if I know what I'm doing. Giving the user an option is nice, but personally I wouldn't force them.

Agree

seabbs commented on R/validate.R:
```diff
@@ -18,16 +18,84 @@
+  assert_subset(sample_id, names(data), empty.ok = TRUE)
+
+  # rename columns
+  if (!is.null(observed)) {
```
rather than doing it this way I was sort of assuming you would have the default be `observed = "observed"` etc and then rename by selection? Is there a reason you went this way?

nikosbosse replied:
Aah.. I just didn't think about it! 

Can do that. Only thing is that there is a bit of a problem with `quantile_level` and `sample_id` - you can't specify both. So either those have to be `NULL` or we'd need to check for the existence of the column.

nikosbosse replied:
I like the mechanic - I'm not sure I like the kinds of error messages it produces. The following is the result if we do an input check such as `assert_subset(predicted, names(data))`, just like before:
<img width="1274" alt="image" src="https://github.com/epiforecasts/scoringutils/assets/37978797/2f436ddd-4eca-4d76-a439-da84f3d083b3">

We can put in some extra work to make the message a bit prettier, but it doesn't strike me as very elegant either. The fact that we're always trying to rename something, even if the user didn't consciously specify an argument makes it a bit hard to come up with a good error message if the default arguments don't match what the user has. 
<img width="558" alt="image" src="https://github.com/epiforecasts/scoringutils/assets/37978797/2dde6448-b730-4e02-aeb8-8468e33faa36">

In terms of the error messages the "only rename something if the user actively specifies it and then validate the end result" seems more elegant. The error messages you would have gotten earlier when attempting to run something like the following
<img width="597" alt="image" src="https://github.com/epiforecasts/scoringutils/assets/37978797/024630a4-cb08-4686-9741-196237704d6e">
would have been `Error: Assertion on 'data' failed: Both columns `observed` and predicted` are needed.` which seems clearer to me.

seabbs reviewed (approved):
Aside from my comment about the default args I like this as is. Happy to watch the discussion play out.

nikosbosse commented:
Ha fun thing. I think `get_forecast_unit()` can't ever fail at the moment... It relies on `colnames()` but `colnames()` simplry returns `NULL` and never errors. I updated the code to reflect that. 

@sbfnk I think you need to approve the suggested changes.

toshiakiasakura commented:
> But as a user, I think I wouldn't like this as it forces me to do a lot of extra typing even if I know what I'm doing. Giving the user an option is nice, but personally I wouldn't force them.

Agree, but is it a user-side problem? If you don't want to type a lot, just write a simple wrapper by yourself in your application. I often do so. As a package, I prefer to follow the principle of the least surprise. What do you think about this point? 
Sorry to bother you, but this is my last challenge!

nikosbosse commented:
> > But as a user, I think I wouldn't like this as it forces me to do a lot of extra typing even if I know what I'm doing. Giving the user an option is nice, but personally I wouldn't force them.
> 
> Agree, but is it a user-side problem? If you don't want to type a lot, just write a simple wrapper by yourself in your application. I often do so. As a package, I prefer to follow the principle of the least surprise. What do you think about this point? Sorry to bother you, but this is my last challenge!

I think we should not assume that all our users would want to do that. I see where you're coming from (and thank you for arguing your point!), but I still think it would be good to make things as easy and as convenient as possible for users. As long as the validation clearly fails when something goes wrong, I think all is fine. Your argument works in reverse as well: if a sophisticated user prefers a different way to input data, they can also write a wrapper.

nikosbosse commented on R/validate.R:
```diff
@@ -18,16 +18,84 @@
+                                forecast_unit = NULL,
+                                quantile_level = NULL,
+                                sample_id = NULL,
+                                ...) {
```
Ah those are good points!

nikosbosse replied:
I think `as_forecast.default()` needs `...` because the generic has it and they always need to have the same arguments. But I like the idea of having a `forecast_type` argument

seabbs replied:
Is that true for `...` args though? I'm not sure it is.

nikosbosse replied:
Just checked, it's true for `...` args as well: 

```
checking S3 generic/method consistency (1.7s)
   as_forecast:
     function(data, ...)
   as_forecast.default:
     function(data, forecast_unit, forecast_type, observed, predicted,
              model, quantile_level, sample_id)
   See section ‘Generic functions and methods’ in the ‘Writing R
   Extensions’ manual.   
```

nikosbosse replied:
Hm. Since additional arguments just have no effect at all, I would personally not throw an error for this. Given that the `...` are there I would not expect an error as a user if I provide an additional argument.

nikosbosse commented:
I just pushed a new commit that updates the documentation a bit. 

Open questions (either before merging this PR or for a new PR): 
- [ ] should `as_forecast()` have a `forecast_type` argument? This could be used to warn the user if their forecast type does not match the one they want. In the case of binary/point this may be useful in particular
- [ ] do we want the arguments to be `observed = NULL` or `observed = "observed"` (see discussion above)

nikosbosse commented:
Perfect. I made the following updates: 
- `as_forecast()` now has an argument `forecast_type`
- added documentation and updated tests for that

Also note that compared to the very first proposal, I changed the order of the arguments, which now is 
```
as_forecast.default <- function(data,
                                forecast_unit = NULL,
                                forecast_type = NULL,
                                observed = NULL,
                                predicted = NULL,
                                model = NULL,
                                quantile_level = NULL,
                                sample_id = NULL,
                                ...) {
```
It felt more natural that way as `forecast_unit()` and `forecast_type()` are the ones you should use almost every time + then we have all arguments related to renaming stuff lumped together.

nikosbosse commented:
Unsure where the failing snapshots on `macOS-latest` are coming from. Can't reproduce that locally...

nikosbosse commented:
Update: given that `macOS-latest` is failing on `main` as well I'd be tempted to merge in this PR and #633 regardless... 
That would allow me to keep developing :) 

<img width="462" alt="image" src="https://github.com/epiforecasts/scoringutils/assets/37978797/0b187bea-924b-49d6-9be2-872eadb7b603">

seabbs commented:
This snapshot failure is due to the recent `ggplot2` update. I think we can ignore it.

seabbs commented on R/validate.R:
```diff
@@ -18,19 +28,104 @@
 #' @export
-as_forecast.default <- function(data, ...) {
+as_forecast.default <- function(data,
+                                forecast_unit = NULL,
```
As a final check why are we using `NULL` defaults vs missing? It seems like a lot of repeated text for no reason.

nikosbosse commented:
Are we happy to merge as is (bypassing branch protection)?
